sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala

```diff
@@ -128,6 +129,45 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
       TableIdentifier("tempTable"), ignoreIfNotExists = true, purge = false)
   }
 
+  test("SPARK-21079 - analyze table with location different than that of individual partitions") {
+    def queryTotalSize(tableName: String): BigInt =
+      spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes
+
+    val tableName = "analyzeTable_part"
+    withTable(tableName) {
+      withTempPaths(4) {
+        case tablePath :: partitionPaths =>
+          sql(
+            s"""
+               |CREATE TABLE ${tableName} (key STRING, value STRING) PARTITIONED BY (ds STRING)
+               |LOCATION '${tablePath}'
+             """.
+              stripMargin).collect()
+
+          val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
+          partitionDates.zip(partitionPaths).foreach {
+            case (ds, path) =>
+              sql(
+                s"""
+                   |ALTER TABLE ${tableName} ADD PARTITION (ds='${ds}')
+                   |LOCATION '${path.toString}'
+                """.
+                  stripMargin).collect()
+              sql(
+                s"""
+                   |INSERT INTO TABLE ${tableName} PARTITION (ds='${ds}')
+                   |SELECT * FROM src
+                """.
+                  stripMargin).collect()
+          }
+
+          sql(s"ANALYZE TABLE ${tableName} COMPUTE STATISTICS noscan")
+
+          assert(queryTotalSize(tableName) === BigInt(17436))
```

Comment:
It can be simplified to 
```Scala
    val tableName = "analyzeTable_part"
    withTable(tableName) {
      withTempPath { path =>
        sql(
          s"""
             |CREATE TABLE $tableName (key STRING, value STRING) PARTITIONED BY (ds STRING)
             |LOCATION '$path'
           """.
            stripMargin)

        val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
        partitionDates.foreach { ds =>
          sql(
            s"""
               |INSERT INTO TABLE $tableName PARTITION (ds='$ds')
               |SELECT * FROM src
             """.stripMargin)
        }

        sql(s"ANALYZE TABLE $tableName COMPUTE STATISTICS noscan")

        assert(
          spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes === BigInt(17436))
      }
    }
```